osgbuild/fetch_sources.py

```diff
@@ -32,194 +82,263 @@
+        log.warning("Absolute path names in .source files break the 4th wall")
+
+    outfile = os.path.join(ops.destdir, os.path.basename(filename or uri))
+    got_sha1sum = download_uri(uri, outfile)
+    log.debug("got sha1sum=%s for uri=%s" % (got_sha1sum, uri))
+
+    if sha1sum: # or not ops.nocheck:
+        check_file_checksum(outfile, sha1sum, got_sha1sum, ops.nocheck)
+
+    return [outfile]
+
+def download_uri(uri, outfile):
+    log.info('Retrieving ' + uri)
+    try:
+        handle = urllib.request.urlopen(uri)
+    except urllib.error.URLError as err:
+        raise Error("Unable to download %s\n%s" % (uri, err))
 
-    tag, tarball = contents.get("tag"), contents.get("tarball")
-    if not tag:
-        msg = "tag not specified: %s" % line
+    sha = hashlib.sha1()
+    try:
+        with open(outfile, 'wb') as desthandle:
+            for chunk in chunked_read(handle):
+                desthandle.write(chunk)
+                sha.update(chunk)
+    except EnvironmentError as e:
+        raise Error("Unable to save downloaded file to %s\n%s" % (outfile, e))
+    return sha.hexdigest()
+
+def chunked_read(handle, size=64*1024):
+    chunk = handle.read(size)
+    while chunk:
+        yield chunk
+        chunk = handle.read(size)
+
+def check_file_checksum(path, sha1sum, got_sha1sum, nocheck):
+    efmt = "sha1 mismatch for '%s':\n    expected: %s\n         got: %s"
+    if sha1sum != got_sha1sum:
+        msg = efmt % (path, sha1sum, got_sha1sum)
         if nocheck:
             log.warning(msg + "\n    (ignored)")
         else:
             raise Error(msg)
-    if tarball:
-        if not tarball[-7:] == ".tar.gz":
-            raise Error("tarball must end with .tar.gz: %s" % line)
-        dest_file = tarball[:-3]  # the .tar file; we'll gzip it ourselves
-    elif tag:
-        tarball_version = tag
-        if re.match("v[0-9]+", tarball_version):
-            tarball_version = tarball_version[1:]
-        # Chop off the release tag (if there is one). This is something like
-        # the "-1" in "3.3-1".  Dashes are not allowed in RPM version numbers
-        # so chop off the first dash and everything afterward.
-        dashidx = tarball_version.find('-')
-        if dashidx != -1:
-            tarball_version = tarball_version[:dashidx]
-        dest_file = "%s-%s.tar" % (name, tarball_version)  # the .tar file; we'll gzip it ourselves
-    else:
-        raise Error("No package tag or tarball specified: %s" % line)
 
-    log.info("Checking out git repo for %s.", name)
+def _required(item, key):
+    if item is None:
+        raise Error("No '%s' specified" % key)
 
-    destdir = os.path.abspath(destdir)
-    checkout_dir = tempfile.mkdtemp(prefix=dest_file, dir=destdir)
-    try:
-        rc = utils.unchecked_call(["git", "clone", git_url, checkout_dir])
-        if rc:
-            raise Error("`git clone %s %s` failed with exit code %d" % (git_url, checkout_dir, rc))
+def _nvl(arg, default):
```

Comment:
what does nvl mean

Reply:
You only use this once! Why are you making me jump around in the code? Just inline this!

Reply:
>what's nvl?

[nvl](https://docs.oracle.com/cd/B19306_01/server.102/b14200/functions105.htm) is an old-time databasey function that provides a default for a null value.  The one-line body `return default if arg is None else arg` is pretty self-explanatory.

>what is a function and why would i want one?

Apparently you do not appreciate the way its one use simplifies and clarifies the code where it is used.  I can inline it if you want but i don't think it will look as pretty then.

Reply:
If it actually simplified and clarified the code then I would not have commented on it. It does not clarify it, it makes it more complicated.